java/code/src/com/redhat/rhn/manager/action/ActionManager.java

```diff
@@ -2009,10 +2009,36 @@ private static List<Map<String, Long>> removeDuplicatedName(List<Map<String, Lon
     public static void addPackageActionDetails(Collection<Action> actions,
             List<Map<String, Long>> packageMaps) {
         if (packageMaps != null) {
+            List<Map<String, Long>> pkgMaps;
+
+            if (actions.iterator().next().getActionType().equals(ActionFactory.TYPE_PACKAGES_REMOVE)) {
+                // our packages.pkgremove state is handling duplicates
+                pkgMaps = packageMaps;
+            }
+            else {
+                long noarch = PackageFactory.lookupPackageArchByLabel("noarch").getId();
+                long all = PackageFactory.lookupPackageArchByLabel("all-deb").getId();
+                Map<String, Map<String, Long>> uPkgMap = new HashMap<>();
+                // For other salt pkg states (pkg.installed), name + arch must be unique.
+                for (Map<String, Long> p : packageMaps) {
+                    long archId = p.getOrDefault("arch_id", noarch);
+                    String name = String.valueOf(p.get("name_id"));
+                    if (archId == noarch || archId == all) {
+                        if (uPkgMap.keySet().stream().noneMatch(k -> k.startsWith(name + "."))) {
+                            uPkgMap.put(name, p);
+                        }
+                    }
+                    else if (!uPkgMap.containsKey(name)) {
+                        String key = name + "." + p.get("arch_id");
+                        uPkgMap.put(key, p);
```

Comment:
I think I'm missing something (maybe because it Friday :) ), but in line 2031 you check map containsKey using just the name, and then adds the key as name+arch. 
I understand that it's looking for duplicate names in arch-specific and also in no-arch (or all) packages. But shouldn't the contains also check if the same key is already inserted?

Reply:
It is a map. Same key would just overwrite the value which is in there. It does not matter which package version is updated, as our state just update the package name and not a specific version.
The arch just matter if `x86_64` and `i586` is installed together. This is how RedHat handle 32Bit packages.

Reply:
ok, got it. Thank you Michael